pkg/p2p/libp2p/libp2p.go

```diff
@@ -238,141 +244,161 @@ func New(ctx context.Context, signer beecrypto.Signer, networkID uint64, overlay
+			s.logger.Warningf("stream handler: don't want incoming peer %s. disconnecting", overlay)
+			_ = handshakeStream.Reset()
+			_ = s.host.Network().ClosePeer(peerID)
 			return
 		}
+	}
 
+	if exists := s.peers.addIfNotExists(stream.Conn(), overlay, i.FullNode); exists {
+		s.logger.Debugf("stream handler: peer %s already exists", overlay)
 		if err = handshakeStream.FullClose(); err != nil {
 			s.logger.Debugf("stream handler: could not close stream %s: %v", overlay, err)
 			s.logger.Errorf("stream handler: unable to handshake with peer %v", overlay)
 			_ = s.Disconnect(overlay)
+		}
+		return
+	}
+
+	if err = handshakeStream.FullClose(); err != nil {
+		s.logger.Debugf("stream handler: could not close stream %s: %v", overlay, err)
+		s.logger.Errorf("stream handler: unable to handshake with peer %v", overlay)
+		_ = s.Disconnect(overlay)
+		return
+	}
+
+	if i.FullNode {
+		err = s.addressbook.Put(i.BzzAddress.Overlay, *i.BzzAddress)
+		if err != nil {
+			s.logger.Debugf("stream handler: addressbook put error %s: %v", peerID, err)
+			s.logger.Errorf("stream handler: unable to persist peer %v", peerID)
+			_ = s.Disconnect(i.BzzAddress.Overlay)
 			return
 		}
+	}
 
-		if i.FullNode {
-			err = s.addressbook.Put(i.BzzAddress.Overlay, *i.BzzAddress)
-			if err != nil {
-				s.logger.Debugf("stream handler: addressbook put error %s: %v", peerID, err)
-				s.logger.Errorf("stream handler: unable to persist peer %v", peerID)
-				_ = s.Disconnect(i.BzzAddress.Overlay)
+	peer := p2p.Peer{Address: overlay, FullNode: i.FullNode}
+
+	s.protocolsmu.RLock()
+	for _, tn := range s.protocols {
+		if tn.ConnectIn != nil {
+			if err := tn.ConnectIn(s.ctx, peer); err != nil {
+				s.logger.Debugf("stream handler: connectIn: protocol: %s, version:%s, peer: %s: %v", tn.Name, tn.Version, overlay, err)
+				_ = s.Disconnect(overlay)
+				s.protocolsmu.RUnlock()
 				return
 			}
 		}
+	}
+	s.protocolsmu.RUnlock()
 
-		peer := p2p.Peer{Address: overlay, FullNode: i.FullNode}
+	if s.notifier != nil {
+		if !i.FullNode {
+			s.lightNodes.Connected(s.ctx, peer)
+			//light node announces explicitly
+			if err := s.notifier.Announce(s.ctx, peer.Address); err != nil {
+				s.logger.Debugf("stream handler: notifier.Announce: %s: %v", peer.Address.String(), err)
+			}
 
-		s.protocolsmu.RLock()
-		for _, tn := range s.protocols {
-			if tn.ConnectIn != nil {
-				if err := tn.ConnectIn(ctx, peer); err != nil {
-					s.logger.Debugf("stream handler: connectIn: protocol: %s, version:%s, peer: %s: %v", tn.Name, tn.Version, overlay, err)
-					_ = s.Disconnect(overlay)
-					s.protocolsmu.RUnlock()
+			if s.lightNodes.Count() > s.lightNodeLimit {
+				// kick another node to fit this one in
+				p, err := s.lightNodes.RandomPeer(peer.Address)
+				if err != nil {
+					s.logger.Debugf("stream handler: cant find a peer slot for light node: %v", err)
+					_ = s.Disconnect(peer.Address)
+					return
+				} else {
+					_ = s.Disconnect(p)
```

Comment:
Can we add a log here?

Reply:
no need to log - this is executed when there's no error